lib/rules/no-wait-for-side-effects.ts

```diff
@@ -0,0 +1,70 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import {
+  getDeepestIdentifierNode,
+  getPropertyIdentifierNode,
+} from '../node-utils';
+import { createTestingLibraryRule } from '../create-testing-library-rule';
+
+export const RULE_NAME = 'no-wait-for-side-effects';
+export type MessageIds = 'noSideEffectsWaitFor';
+type Options = [];
+
+export default createTestingLibraryRule<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: "It's preferred to avoid side effects in `waitFor`",
+      category: 'Best Practices',
+      recommended: false,
+    },
+    messages: {
+      noSideEffectsWaitFor:
+        'Avoid using side effects within `waitFor` callback',
+    },
+    fixable: null,
+    schema: [],
+  },
+  defaultOptions: [],
+  create: function (context, _, helpers) {
+    function hasSideEffects(body: Array<TSESTree.Node>): boolean {
+      return body.some((node: TSESTree.ExpressionStatement) => {
+        const expressionIdentifier = getPropertyIdentifierNode(node);
+
+        if (!expressionIdentifier) {
+          return false;
+        }
+
+        return (
+          helpers.isFireEventUtil(expressionIdentifier) ||
+          helpers.isUserEventUtil(expressionIdentifier)
+        );
+      });
+    }
+
+    function reportSideEffects(node: TSESTree.BlockStatement) {
```

Comment:
Seriously, I can't believe rules can be written so simple now. I want to check a beta of v4 against a real codebase as soon as possible because I can't stop thinking this can't be true and I'll find tons of edge cases which aren't covered by helpers, or everything will explode or something.